prescription/prescription.go

```diff
@@ -0,0 +1,123 @@
+package prescription
+
+import (
+	"context"
+	"time"
+
+	"github.com/globalsign/mgo/bson"
+
+	structureValidator "github.com/tidepool-org/platform/structure/validator"
+	"github.com/tidepool-org/platform/user"
+
+	"github.com/tidepool-org/platform/structure"
+)
+
+const (
+	StateDraft     = "draft"
+	StatePending   = "pending"
+	StateSubmitted = "submitted"
+	StateReviewed  = "reviewed"
+	StateExpired   = "expired"
+	StateActive    = "active"
+	StateInactive  = "inactive"
+
+	MaximumExpirationTime = time.Hour * 24 * 30 // 30 days
+)
+
+type Client interface {
+	Accessor
+}
+
+type Accessor interface {
+	CreatePrescription(ctx context.Context, userID string, create *RevisionCreate) (*Prescription, error)
+}
+
+type Prescription struct {
+	ID               bson.ObjectId `json:"id" bson:"_id"`
+	PatientID        string        `json:"patientId,omitempty" bson:"patientId,omitempty"`
+	AccessCode       string        `json:"accessCode,omitempty" bson:"accessCode"`
+	State            string        `json:"state" bson:"state"`
+	LatestRevision   *Revision     `json:"latestRevision" bson:"latestRevision"`
+	RevisionHistory  Revisions     `json:"-" bson:"revisionHistory"`
+	ExpirationTime   *time.Time    `json:"expirationTime" bson:"expirationTime"`
+	PrescriberUserID string        `json:"prescriberUserId,omitempty" bson:"prescriberUserId,omitempty"`
+	CreatedTime      time.Time     `json:"createdTime" bson:"createdTime"`
+	CreatedUserID    string        `json:"createdUserId" bson:"createdUserId"`
+	DeletedTime      *time.Time    `json:"deletedTime,omitempty" bson:"deletedTime,omitempty"`
+	DeletedUserID    string        `json:"deletedUserId,omitempty" bson:"deletedUserId,omitempty"`
+}
+
+func NewPrescription(userID string, revisionCreate *RevisionCreate) (*Prescription, error) {
+	now := time.Now()
+	accessCode := GenerateAccessCode()
+	revision := NewRevision(userID, 0, revisionCreate)
+	revisionHistory := []*Revision{revision}
+	prescription := &Prescription{
+		ID:               bson.NewObjectId(),
+		AccessCode:       accessCode,
+		State:            revisionCreate.State,
+		LatestRevision:   revision,
+		RevisionHistory:  revisionHistory,
+		ExpirationTime:   revision.CalculateExpirationTime(),
+		CreatedTime:      now,
+		CreatedUserID:    userID,
+		PrescriberUserID: revision.GetPrescriberUserID(),
+	}
+
+	return prescription, nil
+}
+
+type Prescriptions []*Prescription
+
+func (p *Prescription) Validate(validator structure.Validator) {
```

Comment:
That might not be possible for pump settings which is arguably the most important piece of data that needs to be validated. The data service uses union types in some places which are not supported by the Open API Spec libraries for Go. I also noticed that the current spec we have for the data service (pump settings are part of it) is not 100% accurate or is incomplete. Need to do some research whether partial validation is possible.

That being said, I'll try in a separate PR but I see many reasons why it might not work.